graphql_compiler/schema/schema_info.py

```diff
@@ -1,13 +1,166 @@
+    # lead to incorrect output queries being generated.
+    # *****
+    # TODO: make sure we are treating empty dict the same as None
+    type_equivalence_hints: Optional[Dict[str, str]]  # Optional
+
+
+@dataclass
+class MatchSchemaInfo:
+    generic_schema_info: GenericSchemaInfo
+
+
+def create_match_schema_info(
+    schema: GraphQLSchema, type_equivalence_hints: Optional[Dict[str, str]] = None
+) -> MatchSchemaInfo:
+    """Create a SchemaInfo object for a database using MATCH."""
+    generic_schema_info = GenericSchemaInfo(
+        schema=schema, type_equivalence_hints=type_equivalence_hints
+    )
+    return MatchSchemaInfo(generic_schema_info=generic_schema_info)
+
+
+@dataclass
+class GremlinSchemaInfo:
+    generic_schema_info: GenericSchemaInfo
+
+
+def create_gremlin_schema_info(
+    schema: GraphQLSchema, type_equivalence_hints: Optional[Dict[str, str]] = None
+) -> GremlinSchemaInfo:
+    """Create a SchemaInfo object for a database using Gremlin."""
+    generic_schema_info = GenericSchemaInfo(
+        schema=schema, type_equivalence_hints=type_equivalence_hints
+    )
+    return GremlinSchemaInfo(generic_schema_info=generic_schema_info)
+
+
+@dataclass
+class CypherSchemaInfo:
+    generic_schema_info: GenericSchemaInfo
+
+
+def create_cypher_schema_info(
+    schema: GraphQLSchema, type_equivalence_hints: Optional[Dict[str, str]] = None
+) -> CypherSchemaInfo:
+    """Create a SchemaInfo object for a database using Cypher."""
+    generic_schema_info = GenericSchemaInfo(
+        schema=schema, type_equivalence_hints=type_equivalence_hints
+    )
+    return CypherSchemaInfo(generic_schema_info=generic_schema_info)
+
+
+@dataclass
+class SQLSpecificSchemaInfo:
+    # TODO: consider a different name for this class
+    """Schema information specific to SQL databases."""
+
+    # Specifying the dialect for which we are compiling, if a dialect exists
+    # e.g. sqlalchemy.dialects.mssql.dialect()
+    dialect: Dialect
+    # dict mapping every GraphQL object type or interface type name in the schema to
+    # a sqlalchemy table.
+    # Column types that do not exist for this dialect are not allowed.
+    # All tables are expected to have primary keys.
+
+    vertex_name_to_table: Dict[str, sqlalchemy.Table]
+    # dict mapping every GraphQL object type or interface type name in the schema to
+    # dict mapping every vertex field name at that type to a DirectJoinDescriptor.
+    # The tables the join is to be performed on are not specified.
+    # They are inferred from the schema and the tables dictionary.
+    join_descriptors: Dict[str, Dict[str, DirectJoinDescriptor]]
+
+    # TODO: consider extending this to being able to support QueryPlanningSchemaInfo?
+
+
+@dataclass
+class SQLSchemaInfo:
+    # TODO: consider making a class per flavor? Please opine!
+    #       Pros: every flavor has its dedicated suite of classes / functions
+    #       Cons: redundant code + maintenance
```

Comment:
There's some features where our approach could be very different for each dialect. For example inheritance, or list values. So it might be useful to have separate ones in the long run. However, we should have a clear design of these features first, so we can at least know where the difference in compilation would be. Do they all have the same emit function, but different lowering passes?

I think it's too early to tell, and I don't think we'll need such a redesign until version 3.0. I'd leave it for now.

Reply:
Sounds reasonable! I wasn't sure how different they could end up being so this is helpful. I'll adjust the comment.